Branch to Polish UI and Fix stuff after reviews

taa-wtag opened the pull request:
-Made a few changes to UI
-Fixed two bugs
-Added icons and splash
-Changed city mapper extensions
-Fixed Solid Violations

taa-wtag committed:
-Cleaned up Serialized Object to remove unused fields
-Removed DateItem class and flattened hierarchy for CityItem, WeatherHourItem and WeatherDayItem
-Cleaned up RealmDao and removed all Business logic code.
-Added code for duplication check and replace in Dao
-Simplified CityRepository
-Simplified Weather Repository
-Added City Search from Weather Repository as a backup

taa-wtag committed:
-Created multiple Use Cases
-Created Util classes
-Added City ViewModel
-Cleaned up WeatherViewModel

taa-wtag committed:
-Started writing mock Repositories and sometest cases

taa-wtag committed:
-Finished writing all unit tests

taa-wtag committed:
-Finished writing all unit tests

taa-wtag committed:
-Added one missed test case

taa-wtag committed:
-Completed UI and fixed issues related to flow

taa-wtag committed:
-Made a few changes to UI
-Fixed two bugs
-Added icons and splash
-Changed city mapper extensions

ash-wtag commented on app/src/main/java/com/rektstudios/trueweather/data/helper/GeocodeHelper.kt:
```diff
@@ -0,0 +1,21 @@
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            var cityName = ""
+            Geocoder(context, Locale.getDefault()).getFromLocation(lat, lon, 1) {
+                cityName = it[0].locality + ", " + it[0].countryName
```
please add check before accessing index

ash-wtag commented on app/src/main/java/com/rektstudios/trueweather/data/local/RealmDaoImpl.kt:
```diff
@@ -1,115 +1,99 @@
+                                    ?.lessThan("timeEpoch", Calendar.getInstance().timeInMillis/1000L- FORECAST_MIN_TIME_PAST)
+                                    ?.findAll()
+                                    ?.deleteAllFromRealm()
+                            if(weather.timeEpoch>Calendar.getInstance().timeInMillis/1000L- FORECAST_MIN_TIME_PAST)
```
please add proper indentation
please add function for Calendar.getInstance().timeInMillis/1000L and reuse

ash-wtag commented on app/src/main/java/com/rektstudios/trueweather/data/local/RealmDaoImpl.kt:
```diff
@@ -1,115 +1,99 @@
+                                    ?.lessThan("dateEpoch", Calendar.getInstance().timeInMillis/1000L-FORECAST_MIN_TIME_PAST)
+                                    ?.findAll()
+                                    ?.deleteAllFromRealm()
+                            if(weather.dateEpoch>Calendar.getInstance().timeInMillis/1000L- FORECAST_MIN_TIME_PAST)
```
ditto

ash-wtag commented on app/src/main/java/com/rektstudios/trueweather/data/local/RealmDaoImpl.kt:
```diff
@@ -1,115 +1,99 @@
-            realm.copyToRealm(cityItem)
+    override suspend fun addCity(city: String)  {
+        realm.executeTransaction {
+            getCityAsQuery(it, city) ?: run{ it.copyToRealm(CityItem().apply {
```
please add primary key and skip duplicate validation manually

ash-wtag commented on app/src/main/java/com/rektstudios/trueweather/data/location/LocationTrackerImpl.kt:
```diff
@@ -0,0 +1,66 @@
+        }
+    }
+
+    private fun checkLocationPermissionsGranted(): Boolean{
```
please add proper name for the function

ash-wtag commented on app/src/main/java/com/rektstudios/trueweather/data/location/LocationTrackerImpl.kt:
```diff
@@ -0,0 +1,66 @@
+    override suspend fun getCurrentLocation(): Pair<Double, Double>? {
+
+        return if(checkLocationPermissionsGranted()) null
+        else suspendCancellableCoroutine {cont ->
```
please add proper name

ash-wtag commented on app/src/main/java/com/rektstudios/trueweather/data/repository/PrefsRepositoryImpl.kt:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    private suspend fun checkValueExists(key: String){
+        val prefsKey = stringPreferencesKey(key)
```
Please add preference data store keys instead of string

ash-wtag commented on app/src/main/java/com/rektstudios/trueweather/data/repository/PrefsRepositoryImpl.kt:
```diff
@@ -0,0 +1,42 @@
+    private suspend fun checkValueExists(key: String){
+        val prefsKey = stringPreferencesKey(key)
+        when(key){
+            KEY_CITY_NAME -> prefsDataStore.data.firstOrNull()?.get(prefsKey) ?: saveValue(key,"")
```
Please add single responsibility for this function

taa-wtag committed:
-Todo

taa-wtag committed:
-Implemented all review changes
-Converted all realm operation to async
-Converted all realm query items to unmanaged
-Fixed remaining bugs
-Removed unnecessary response objects
-Fixed naming for multiple classes and functions
-Fixed some test cases due to changes in viewmodels

taa-wtag committed:
-Removed unnecessary comment

taa-wtag committed:
-Fixed Formatting using default code reformat

taa-wtag committed:
-Refactored code to remove parts that could cause memory leaks
